README
NAME
    Riak::Client - Fast and lightweight Perl client for Riak

VERSION
    version 1.94

SYNOPSIS
      use Riak::Client;

      # normal mode
      my $client = Riak::Client->new(
        host => '127.0.0.1',
        port => 8087,
        r => 2,
        w => 2,
        dw => 1,
        connection_timeout => 5,
        read_timeout => 5,
        write_timeout => 5,
        no_auto_connect => 0,
      );

      # AnyEvent mode
      my $client = Riak::Client->new(
        host => '127.0.0.1',
        port => 8087
        anyevent_mode => 1,
      );

      $client->is_alive() or die "riak is not alive";

      # store hashref. will be serialized as JSON
      $client->put( 'bucket_name', 'key_name', { some => 'structure' } );

      # store text
      $client->put( 'bucket_name', 'key_name', 'sometext', 'text/plain' );

      # store raw data
      $client->put_raw( 'bucket_name', 'key_name', 'rawdata' );

      # fetch hashref
      my $hash = $client->get( 'bucket_name', 'key_name' );

      # fetch raw data
      my $text = $client->get_raw( 'bucket_name', 'key_name');

      # delete data
      $client->del( 'bucket_name', 'key_name');

      # AnyEvent mode
      my $cv = AE::cv
      $client->get_raw( 'bucket_name', 'key_name'
                        sub { do_something_with($_[0]);
                              $cv->send();
                        }
                  );
      # ... later one
      $cv->recv();

      # list keys in stream
      $client->get_keys(foo => sub{
         my ($key, $done) = @_;

         # you should use another client inside this callback!
         $another_client->del(foo => $key);

      });

DESCRIPTION
    Riak::Client is a fast and light Perl client for Riak using PBC
    interface, with optional AnyEvent mode.

    It supports operations like ping, get, exists, put, del, secondary
    indexes (so-called 2i) setting and querying, and Map Reduce querying.

    It has two modes, a traditional procedural mode, and an event based
    mode, using AnyEvent.

    It started as a fork of "Riak::Light" to fix some bugs, but actually
    ended up in a complete rewrite with more features, but the same
    performance.

ATTRIBUTES
  anyevent_mode
    Enables the AnyEvent mode, allowing true asynchronous mode.

  host
    Str, Required. Riak IP or hostname.

  port
    Int, Required. Port of the PBC interface.

  r
    Int, Default 2. R value setting for this client.

  w
    Int, Default 2. W value setting for this client.

  dw
    Int, Default 1. DW value setting for this client.

  connection_timeout
    Float, Default 5. Timeout for connection operation, in seconds. Set to 0
    for no timeout.

  read_timeout
    Float, Default 5. Timeout for read operation, in seconds. Set to 0 for
    no timeout.

  no_delay
    Boolean, Default 0. If set to a true value, TCP_NODELAY will be enabled
    on the socket, which means deactivating Nagle's algorithm. Use only if
    you know what you're doing.

  no_auto_connect
    Bool, Default 0. If set to true, then the module won't automatically
    connect upon instanciation. Instead, you'll have to call "connect()"
    yourself.

  anyevent_mode
    Bool, Default 0. If set to true, then all methods can receive a
    callback, as last argument. If present, the method will return
    immediately, and the callback will be executed upon completion of the
    operation, receiving a condvar as first and only argument. If set to
    false (the default), then the client instance will be synchronous.

METHODS
  connect
      $client->connect();
      $client->connect($coderef);

    Connects to the Riak server. On error, will raise an exception. This is
    automatically done when "new()" is called, unless the "no_auto_connect"
    attribute is set to true. Accepts an optional callback, that will be
    executed when connected.

      # example in AnyEvent mode
      $cv = AE::cv;
      $client->connect(sub { print "connected!\n"; $cv->send(); });
      # ...
      $cv->recv();

  ping
     my $result = $client->ping();
     $client->ping($coderef);

    Performs a ping operation. On error, will raise an exception. Accepts an
    optional callback, that will be executed upon completion

      # example in AnyEvent mode
      $cv = AE::cv;
      $client->ping(sub { print "got $_[0] \n"; $cv->send(); });
      # ...
      $cv->recv();

      # an other example
      use Try::Tiny;
      try { $client->ping() } catch { "oops... something is wrong: $_" };

    See also "is_alive()".

  is_alive
     my $is_alive = $client->is_alive();
     $client->is_alive($coderef);

    Checks if the connection is alive. Returns true or false. On error, will
    raise an exception. Accepts an optional callback, that will be executed
    upon completion. Even in AnyEvent mode, this operation is synchronous.

      # example in AnyEvent mode
      $cv = AE::cv;
      $client->is_alive(sub { print($_[0] ? "alive\n" : "dead\n"); $cv->send(); });
      # ...
      $cv->recv();

  get
      my $value = $client->get($bucket, $key);
      $client->get($bucket, $key, $coderef);

      # example in AnyEvent mode
      $cv = AE::cv;
      $client->get('bucket', 'key', sub { do_stuff_with_value($_[0]); $cv->send() });
      # ...
      $cv->recv();

    Performs a fetch operation. Expects bucket and key names. Returns the
    value. On error, will raise an exception. Accepts an optional callback,
    that will be called upon completion, with the value as first argument.
    If the content_type of the fetched value is 'application/json',
    automatically decodes the JSON into a Perl structure. If you need the
    raw data you can use "get_raw".

  get_raw
      my $value = $client->get_raw($bucket, $key);
      $client->get_raw($bucket, $key, $coderef);

    Same as "get", but no automatic JSON decoding will be performed. If you
    want JSON to be automatically decoded, you should use "get()" instead.

  put
      $client->put($bucket, $key, $value);
      $client->put($bucket, $key, $value, $coderef);
      $client->put($bucket, $key, $value, $mime_type, $coderef);
      $client->put($bucket, $key, $value, $mime_type, $secondary_indexes, $coderef);
      $client->put($bucket, $key, $value, $mime_type, $secondary_indexes, $links, $coderef);

    Performs a store operation. Expects bucket and key names, the value, the
    content type (optional, default is 'application/json'), the indexes to
    set for this value (optional, default is none), the links to set for
    this value (optional, default is none), and an optional coderef. On
    error, will raise an exception

    Will encode the structure in json string if necessary. If you need to
    store the raw data you should use "put_raw" instead.

    IMPORTANT: all the index field names should end by either "_int" or
    "_bin", depending if the index type is integer or binary.

    To query secondary indexes, see "query_index".

      $client->put('bucket', 'key', { some_values => [1,2,3] });
      $client->put('bucket', 'key', { some_values => [1,2,3] }, 'application/json);
      $client->put('bucket', 'key', 'text', 'plain/text');

      # you can set secondary indexes (2i)
      $client->put( 'bucket', 'key', 'text_value', 'plain/text',
                    { field1_bin => 'abc', field2_int => 42 }
                  );
      $client->put( 'bucket', 'key', { some_values => [1,2,3] }, undef,
                    { field1_bin => 'abc', field2_int => 42 }
                  );

      # you can also set links
      $client->put( 'bucket', 'key', 'text', 'plain/text', undef,
                    { link_tag1 => 'bucket/key',
                      link_tag2 => 'other_bucket/key',
                    }
                  );

      # you can set multiple links for the same tag
      $client->put( 'bucket', 'key', 'text', 'plain/text', undef,
                    { link_tag1 => [ qw( bucket/key bucket2/key2 ) ],
                      link_tag2 => 'other_bucket/key',
                    }
                  );

      # you can also use this form (marginally faster)
      $client->put( 'bucket', 'key', 'text', 'plain/text', undef,
                    [ { tag => 'link_tag1', bucket => 'bucket1', key => 'key1'},
                      { tag => 'link_tag2', bucket => 'bucket2', key => 'key2'},
                    ],
                  );

      # example in AnyEvent mode
      $cv = AE::cv;
      $client->put( 'bucket', 'key', 'some_text', 'plain/text',
                    { field1_bin => 'abc', field2_int => 42 },
                    { next_key => 'bucket2/foo'},
                    sub { print "data is sent to Riak\n"; $cv->send() },
                  );
      # ...
      $cv->recv();

  put_raw
      $client->put_raw('bucket', 'key', encode_json({ some_values => [1,2,3] }), 'application/json');
      $client->put_raw('bucket', 'key', 'text');
      $client->put_raw('bucket', 'key', 'text', undef, {field_bin => 'foo'});
      $client->put_raw('bucket', 'key', 'text', undef, {field_bin => 'foo'}, $links);

    For more example, see "put".

    Perform a store operation. Expects bucket and key names, the value, the
    content type (optional, default is 'plain/text'), the indexes (optional,
    default is none), and links (optional, default is none) to set for this
    value

    This method won't encode the data, but pass it as such, trusting it's in
    the type you've indicated in the passed content-type. If you want the
    structure to be automatically encoded, use "put" instead.

    IMPORTANT: all the index field names should end by either "_int" or
    "_bin", depending if the index type is integer or binary.

    To query secondary indexes, see "query_index".

  del
      $client->del(bucket => key);

    Perform a delete operation. Expects bucket and key names.

  get_keys
      # in default mode
      $client->get_keys(foo => sub{
         my ($key, $done) = @_;
         # you should use another client inside this callback!
         $another_client->del(foo => $key);
      });

      # in anyevent mode
      my $cv = AE::cv;
      $client->get_keys(foo => sub{
         my ($key, $done) = @_;
         # ... do stuff with $key
         $done and $cv->send;
      });
      $cv->recv();

    WARNING, this method should not be called on a production Riak cluster,
    as it can have a big performance impact. See Riak's documentation.

    WARNING, because Riak doesn't handles pipelining, you cannot use the
    same "Riak::Client" instance inside the callback, it would raise an
    exception.

    Perform a list keys operation. Receive a callback and will call it for
    each key. The callback will receive two arguments: the key, and a
    boolean indicating if it's the last key

    The callback is optional, in which case an ArrayRef of all the keys are
    returned. But don't do that, and always provide a callback, to avoid
    your RAM usage to skyrocket...

  exists
      $client->exists(bucket => 'key') or warn "key not found";

    Perform a fetch operation but with head => 0, and the if there is
    something stored in the bucket/key.

  query_index
    Perform a secondary index (2i) query. Expects a bucket name, the index
    field name, the index value you're searching on, and optionally a
    callback.

    If a callback has been provided, doesn't return anything, but execute
    the callback on each matching keys. callback will receive the key name

Changes
1.94      2014-07-23 14:29:18 Europe/Amsterdam

   * more mem leaks fix, documentation, and better exception message (and trace)

1.93      2014-07-22 15:05:51 Europe/Amsterdam

   * chase down memory leaks (Ivan + dams)

1.92      2014-07-12 23:20:44 Europe/Amsterdam

   * fix requires

1.91      2014-07-11 11:41:55 Europe/Amsterdam

   * faster AnyEvent mode
   * supports Links setting, Link Walking, etc.
   * minor fixes.
   * better documentation

1.90      2014-06-29 23:11:22 Europe/Amsterdam

   * complete rewrite from scratch
   * AnyEvent mode
   * callback mode
   * supports Map Reduce
   * bugfixes, documentation, tests

0.11      2013-11-21 14:16:06 Europe/Amsterdam
   * fix networking code issues (handle EINTR and fix bug) (ivan)
   * implements get_buckets + (get|set)_bucket_props (ivan)
   * other bugfixes

0.10      2013-11-20 17:29:38 Europe/Amsterdam
   * almost rewrote it from scratch (dams)
   * a lot of refactoring (dams)
   * removed the autodie option (dams)
   * added pbc protocol (dams)
   * better dist.ini (dams)
   * Initial version, forked from Riak::Light (dams)


lib/Riak/Client/PBC.pm
#
# This file is part of Riak-Client
#
# This software is copyright (c) 2014 by Damien Krotkine.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Riak::Client::PBC;
{
  $Riak::Client::PBC::VERSION = '1.94';
}

##
## This file was generated by Google::ProtocolBuffers (0.09)
##  on Sun Jun 29 15:03:36 2014 from file /Users/dkrotkine/Devel/riak-client/pbc/riak.proto
##

use strict;
use warnings;
use Google::ProtocolBuffers;
{
    unless (RpbIndexReq::IndexQueryType->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_enum(
            'RpbIndexReq::IndexQueryType',
            [
               ['eq', 0],
               ['range', 1],

            ]
        );
    }
    
    unless (RpbIndexResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbIndexResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'keys', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'results', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'continuation', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 4, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbIndexReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbIndexReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'index', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbIndexReq::IndexQueryType', 
                    'qtype', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'range_min', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'range_max', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_terms', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'stream', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'max_results', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'continuation', 10, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbContent', 
                    'content', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'unchanged', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbBucketProps->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbBucketProps',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'n_val', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'allow_mult', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'r', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pr', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'basic_quorum', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'notfound_ok', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'if_modified', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'head', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'deletedvclock', 9, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPutReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPutReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbContent', 
                    'content', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'w', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'dw', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_body', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pw', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'if_not_modified', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'if_none_match', 10, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'return_head', 11, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPutResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPutResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbContent', 
                    'contents', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetBucketReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetBucketReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbListKeysResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbListKeysResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'keys', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbPair->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbPair',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'value', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbListKeysReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbListKeysReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbGetBucketResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbGetBucketResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbBucketProps', 
                    'props', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbDelReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbDelReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'rw', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vclock', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'r', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'w', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pr', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'pw', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'dw', 9, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbContent->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbContent',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'value', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_type', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'charset', 3, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_encoding', 4, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'vtag', 5, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbLink', 
                    'links', 6, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'last_mod', 7, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'last_mod_usecs', 8, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'usermeta', 9, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    'RpbPair', 
                    'indexes', 10, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'deleted', 11, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbSetBucketReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbSetBucketReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    'RpbBucketProps', 
                    'props', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbMapRedResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbMapRedResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'phase', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'response', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BOOL(), 
                    'done', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbErrorResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbErrorResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'errmsg', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_UINT32(), 
                    'errcode', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbMapRedReq->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbMapRedReq',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'request', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_REQUIRED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'content_type', 2, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbLink->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbLink',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'bucket', 1, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'key', 2, undef
                ],
                [
                    Google::ProtocolBuffers::Constants::LABEL_OPTIONAL(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'tag', 3, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

    unless (RpbListBucketsResp->can('_pb_fields_list')) {
        Google::ProtocolBuffers->create_message(
            'RpbListBucketsResp',
            [
                [
                    Google::ProtocolBuffers::Constants::LABEL_REPEATED(), 
                    Google::ProtocolBuffers::Constants::TYPE_BYTES(), 
                    'buckets', 1, undef
                ],

            ],
            { 'create_accessors' => 1,  }
        );
    }

}
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Riak::Client::PBC

=head1 VERSION

version 1.94

=head1 AUTHOR

Damien Krotkine <dams@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Damien Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
